examples/cryptocurrency-advanced/frontend/src/plugins/blockchain.js

```diff
@@ -4,200 +4,142 @@ import * as proto from '../../proto/stubs.js'
-                  // to number of transactions in array with transactions meta data
-                  throw new Error('Transactions can not be verified')
-                }
-
-                // validate each transaction
-                const transactions = []
-                let index = 0
-
-                for (let transaction of data.wallet_history.transactions) {
-                  const hash = transactionsMetaData[index++]
-                  const buffer = Exonum.hexadecimalToUint8Array(transaction.message)
-                  const bufferWithoutSignature = buffer.subarray(0, buffer.length - 64)
-                  const author = Exonum.uint8ArrayToHexadecimal(buffer.subarray(0, 32))
-                  const signature = Exonum.uint8ArrayToHexadecimal(buffer.subarray(buffer.length - 64, buffer.length));
-
-                  const Transaction = getTransaction(transaction.debug, author)
-
-                  if (Exonum.hash(buffer) !== hash) {
-                    throw new Error('Invalid transaction hash')
-                  }
-
-                  // serialize transaction and compare with message
-                  if (!Transaction.serialize(transaction.debug).every(function (el, i) {
-                    return el === bufferWithoutSignature[i]
-                  })) {
-                    throw new Error('Invalid transaction message')
-                  }
-
-                  if (!Transaction.verifySignature(signature, author, transaction.debug)) {
-                    throw new Error('Invalid transaction signature')
-                  }
-
-                  const transactionData = Object.assign({ hash: hash }, transaction.debug)
-                  if (transactionData.to) {
-                    transactionData.to = Exonum.uint8ArrayToHexadecimal(new Uint8Array(transactionData.to.data))
-                  }
-                  transactions.push(transactionData)
-                }
-
-                return {
-                  block: data.block_proof.block,
-                  wallet: wallet,
-                  transactions: transactions
-                }
-              })
+            .then(({ block_proof, wallet_proof, wallet_history }) => {
+              if (Exonum.verifyBlock(block_proof, validators) !== undefined) throw new Error('Block is invalid')
+              const tableRootHash = Exonum.verifyTable(wallet_proof.to_table, block_proof.block.state_hash, 'crypto.wallets')
+              const walletProof = new Exonum.MapProof(wallet_proof.to_wallet, Exonum.MapProof.rawKey(Exonum.PublicKey), Wallet)
+              if (walletProof.merkleRoot !== tableRootHash) throw new Error('Wallet proof is corrupted')
+
+              const wallet = walletProof.entries.get(Exonum.publicKeyToAddress(publicKey))
+              if (typeof wallet === undefined) throw new Error('Wallet not found')
+
+              const transactions = wallet_history.transactions.map(deserializeWalletTx)
```

Comment:
You've skipped some proof checks, e.g., that transaction hashes are appropriate. See exonum/exonum-doc#394 for how I see these checks.

Reply:
I did :)